### PR TITLE
🎉  Source Harvest: add OAuth 2.0

### LINKED ISSUE #7707: Source Harvest: support oauth
## Tell us about the problem you're trying to solve
With the release of Airbyte Cloud, we need to start supporting Oauth for this connector, since it's the recommended way of authenticating users into a SaaS application.

If this connector doesn't support oauth already (i.e: doesn't accept a client_id and client_secret) then we need to update its spec to accept those parameters. There are two ways to do this:

If the connector already supports some auth mechanism like api_key, I suggest that this be a oneof nested inside a top-level field called "authentication":

{ authentication: { type: object oneOf: [ // api key, // oauth ] } }

If the connector only supports webflow oauth, then no changes are needed to the properties format and we will only need to add annotations.

See the connector spec reference in the docs for reference on how a oneof can be implemented.

This should be done in a backwards compatible manner i.e: users currently supplying authentication info in the config's top-level should not be impacted by this change.
## Acceptance criteria
The connector supports oauth webflow authentication with client_id/client_secret in a backwards compatible manner
Oauth properties are annotated properly. See this PR for an example

### PR DESCRIPTION
**Pull request by grubberr**
Signed-off-by: Sergey Chvalyuk <grubberr@gmail.com>

## How
Add an implementation of oAuth2 logic to the connector and the Java server part

Manual result of tests:

1. Redirect to the Harvest auth window:
Screenshot 2021-11-16 at 20 59 22
2. Auth result:
Screenshot 2021-11-16 at 21 01 19

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

### CODE REVIEW

**Review by bazarnov — APPROVED** [review]

File: airbyte-integrations/connectors/source-harvest/source_harvest/auth.py [hunk 1]
```diff
@@ -4,14 +4,22 @@
...
-from airbyte_cdk.sources.streams.http.auth import TokenAuthenticator
+from airbyte_cdk.sources.streams.http.auth import Oauth2Authenticator, TokenAuthenticator
 
 
-class HarvestTokenAuthenticator(TokenAuthenticator):
-    def __init__(self, token: str, account_id: str, account_id_header: str = "Harvest-Account-ID", **kwargs):
-        super().__init__(token, **kwargs)
+class HarvestMixin:
+    def __init__(self, *, account_id: str, account_id_header: str = "Harvest-Account-ID", **kwargs):
+        super().__init__(**kwargs)
```
Comment (antixar) [inline airbyte-integrations/connectors/source-harvest/source_harvest/auth.py:12]: what is the simple '*'?

File: airbyte-integrations/connectors/source-harvest/source_harvest/auth.py [hunk 2]
```diff
@@ -4,14 +4,22 @@
...
     def get_auth_header(self) -> Mapping[str, Any]:
         return {**super().get_auth_header(), self.account_id_header: self.account_id}
+
+
+class HarvestTokenAuthenticator(HarvestMixin, TokenAuthenticator):
+    pass
+
+
+class HarvestOauth2Authenticator(HarvestMixin, Oauth2Authenticator):
+    pass
```
Comment (antixar) [inline airbyte-integrations/connectors/source-harvest/source_harvest/auth.py:25]: Please add code comments about goals of this classes

**Review by antixar — CHANGES_REQUESTED** [review]
Please add basic tests for Java part. You can use examples of other ready connectors.

File: airbyte-integrations/connectors/source-harvest/source_harvest/auth.py (hunk 2 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-harvest/source_harvest/auth.py]: I have added docs for auth classes

File: airbyte-integrations/connectors/source-harvest/source_harvest/auth.py (hunk 1 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-harvest/source_harvest/auth.py:16]: PEP 3102 ?

Comment (antixar) [reply airbyte-integrations/connectors/source-harvest/source_harvest/auth.py:16]: You are right. it's good

**Review by antixar — APPROVED** [review]
Looks good!

**Review by bazarnov — APPROVED** [review]

File: airbyte-integrations/connectors/source-harvest/source_harvest/spec.json [hunk 3]
```diff
@@ -25,9 +19,77 @@
...
-  "supported_destination_sync_modes": ["append"]
+  "supported_destination_sync_modes": ["append"],
+  "authSpecification": {
+    "auth_type": "oauth2.0",
+    "oauth2Specification": {
+      "rootObject": ["credentials", 0],
+      "oauthFlowInitParameters": [["client_id"], ["client_secret"]],
+      "oauthFlowOutputParameters": [["refresh_token"]]
+    }
+  }
```
Comment (ChristopheDuong) [inline airbyte-integrations/connectors/source-harvest/source_harvest/spec.json:94]: Could we start adding the new way of specifying OAuth with `advanced_auth` objects since #7827 is now live?

We'll be deprecating this "authSpecification" in the spec.json afterward

Comment (grubberr) [reply airbyte-integrations/connectors/source-harvest/source_harvest/spec.json]: I have added "advancedAuth" section to spec.json

File: airbyte-integrations/connectors/source-harvest/source_harvest/spec.json [hunk 4]
```diff
@@ -25,9 +19,106 @@
...
+          "type": "string"
+        },
+        "client_secret": {
+          "type": "string"
+        }
+      },
+      "complete_oauth_server_output_specification": {
+        "client_id": {
+          "type": "string",
+          "path_in_connector_config": ["credentials", 0, "client_id"]
```
Comment (ChristopheDuong) [inline airbyte-integrations/connectors/source-harvest/source_harvest/spec.json:115]:
```suggestion
          "path_in_connector_config": ["credentials", "client_id"]
```

File: airbyte-integrations/connectors/source-harvest/source_harvest/spec.json [hunk 5]
```diff
@@ -25,9 +19,106 @@
...
+        },
+        "client_secret": {
+          "type": "string",
+          "path_in_connector_config": ["credentials", 0, "client_secret"]
```
Comment (ChristopheDuong) [inline airbyte-integrations/connectors/source-harvest/source_harvest/spec.json:119]:
```suggestion
          "path_in_connector_config": ["credentials", "client_secret"]
```

File: airbyte-integrations/connectors/source-harvest/source_harvest/spec.json [hunk 6]
```diff
@@ -25,9 +19,106 @@
...
+    "oauth2Specification": {
+      "rootObject": ["credentials", 0],
+      "oauthFlowInitParameters": [["client_id"], ["client_secret"]],
+      "oauthFlowOutputParameters": [["refresh_token"]]
+    }
+  },
+  "advancedAuth": {
+    "auth_flow_type": "oauth2.0",
+    "oauth_config_specification": {
+      "complete_oauth_output_specification": {
```
Comment (ChristopheDuong) [inline airbyte-integrations/connectors/source-harvest/source_harvest/spec.json:98]:
```suggestion
    "oauth_config_specification": {
      "predicate_key": ["credentials", "auth_type"],
      "predicate_value": "Client",      
      "complete_oauth_output_specification": {
```

**Review by ChristopheDuong — CHANGES_REQUESTED** [review]
The object inside:
- complete_oauth_output_specification
- complete_oauth_server_input_specification
- complete_oauth_server_output_specification

needs to be JSON schema object, so you need to add a layer of 
```
 "type": "object",
 "additionalProperties": false,
 "properties": {
```

File: airbyte-integrations/connectors/source-harvest/source_harvest/spec.json [hunk 7]
```diff
@@ -25,9 +19,106 @@
...
+        }
+      },
+      "complete_oauth_server_input_specification": {
+        "client_id": {
+          "type": "string"
+        },
+        "client_secret": {
+          "type": "string"
+        }
+      },
```
Comment (ChristopheDuong) [inline airbyte-integrations/connectors/source-harvest/source_harvest/spec.json:111]:
```suggestion
      "complete_oauth_server_input_specification": {
        "type": "object",
        "additionalProperties": "false",
        "properties": {
          "client_id": {
            "type": "string"
          },
          "client_secret": {
            "type": "string"
          }
        }
      },
```

File: airbyte-integrations/connectors/source-harvest/source_harvest/spec.json (hunk 6 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-harvest/source_harvest/spec.json:98]: is it mistake here ?
I see that `predicate_key` not inside `oauth_config_specification`

https://github.com/airbytehq/airbyte/blob/c5a726737891b5802eaf1515b8503f75eb56bbcb/airbyte-api/src/main/openapi/config.yaml#L3255

Comment (ChristopheDuong) [reply airbyte-integrations/connectors/source-harvest/source_harvest/spec.json:98]: oh you're right then it's:

```suggestion
    "predicate_key": ["credentials", "auth_type"],
    "predicate_value": "Client",      
    "oauth_config_specification": {
      "complete_oauth_output_specification": {
```

File: airbyte-integrations/connectors/source-harvest/source_harvest/spec.json [hunk 8]
```diff
@@ -25,9 +19,106 @@
...
+      "oauthFlowOutputParameters": [["refresh_token"]]
+    }
+  },
+  "advancedAuth": {
+    "auth_flow_type": "oauth2.0",
+    "oauth_config_specification": {
+      "complete_oauth_output_specification": {
+        "refresh_token": {
+          "type": "string",
+          "path_in_connector_config": ["credentials", 0, "refresh_token"]
```
[the remark this answers is not in the document]
Comment (grubberr) [reply airbyte-integrations/connectors/source-harvest/source_harvest/spec.json]: fixed

File: airbyte-integrations/connectors/source-harvest/source_harvest/spec.json (hunk 4 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-harvest/source_harvest/spec.json]: fixed

File: airbyte-integrations/connectors/source-harvest/source_harvest/spec.json (hunk 5 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-harvest/source_harvest/spec.json]: fixed

File: airbyte-integrations/connectors/source-harvest/source_harvest/spec.json (hunk 6 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-harvest/source_harvest/spec.json:100]: fixed

File: airbyte-integrations/connectors/source-harvest/source_harvest/spec.json (hunk 7 above)
Comment (grubberr) [reply airbyte-integrations/connectors/source-harvest/source_harvest/spec.json]: fixed

**Review by ChristopheDuong — APPROVED** [review]

File: tools/bin/ci_credentials.sh [hunk 9]
```diff
@@ -224,7 +224,6 @@ read_secrets source-google-sheets "$GOOGLE_SHEETS_TESTS_CREDS_OLD" "old_config.j
 read_secrets source-google-workspace-admin-reports "$GOOGLE_WORKSPACE_ADMIN_REPORTS_TEST_CREDS"
 read_secrets source-greenhouse "$GREENHOUSE_TEST_CREDS"
 read_secrets source-greenhouse "$GREENHOUSE_TEST_CREDS_LIMITED" "config_users_only.json"
-read_secrets source-harvest "$HARVEST_INTEGRATION_TESTS_CREDS"
```
Comment (ChristopheDuong) [inline tools/bin/ci_credentials.sh:227]: curious why are these removed?

Comment (grubberr) [reply tools/bin/ci_credentials.sh:227]: We are switching from Github secrets to Google Secrets. New approach based on google secrets allow to dynamically generate secrets/config.json for current connector without this hard-coded variable.
